signer/awsv2/sdkv2signer.go

```diff
@@ -0,0 +1,106 @@
+// SPDX-License-Identifier: Apache-2.0
+//
+// The OpenSearch Contributors require contributions made to
+// this file be licensed under the Apache-2.0 license or a
+// compatible open source license.
+//
+// Modifications Copyright OpenSearch Contributors. See
+// GitHub history for details.
+
+package awsv2
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"github.com/aws/aws-sdk-go-v2/aws"
+	awsSignerV4 "github.com/aws/aws-sdk-go-v2/aws/signer/v4"
+	"github.com/opensearch-project/opensearch-go/v2/signer"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"time"
+)
+
+const (
+	openSearchService = "es"
+	emptyStringSHA256 = `e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855`
+)
+
+type awsSdkV2Signer struct {
+	service string
+	signer  *awsSignerV4.Signer
+	awsCfg  aws.Config
+}
+
+// NewSigner returns an instance of Signer for AWS OpenSearchService
+func NewSigner(cfg aws.Config) (signer.Signer, error) {
+	return NewSignerWithService(cfg, openSearchService)
+}
+
+// NewSignerWithService returns an instance of Signer for given service
+func NewSignerWithService(cfg aws.Config, service string) (signer.Signer, error) {
+	if len(strings.TrimSpace(service)) < 1 {
+		return nil, errors.New("service cannot be empty")
+	}
+
+	return &awsSdkV2Signer{
+		service: service,
+		signer:  awsSignerV4.NewSigner(),
+		awsCfg:  cfg,
+	}, nil
+}
+
+func (s *awsSdkV2Signer) SignRequest(r *http.Request) error {
+	ctx := context.Background()
+	t := time.Now()
+
+	creds, err := s.awsCfg.Credentials.Retrieve(ctx)
+	if err != nil {
+		return err
+	}
+
+	if len(s.awsCfg.Region) == 0 {
+		return fmt.Errorf("aws region cannot be empty")
+	}
+
+	hash := emptyStringSHA256
+	if r.Body != nil {
+		hash, err = hexEncodedSha256OfRequest(r)
+		if err != nil {
+			return err
+		}
+
+		// AWS SDK v2 signer will append the content length value twice
+		// and produce an incorrect signature if previously set.
+		// To avoid this, we have to remove the content-length header.
+		r.Header.Del("Content-Length")
+	}
+
+	if err := s.signer.SignHTTP(ctx, creds, r, hash, s.service, s.awsCfg.Region, t); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func hexEncodedSha256OfRequest(r *http.Request) (string, error) {
```

Comment:
How about checking if request body == nil return empty hash or return calculated hash value from this method. In that way, hashing logic is within single method. WDYT?

Reply:
I tried to keep the code as close to the original AWS SDK v1 Signer implementation as possible. In AWS SDK V1 the signer is the one calculating the body hash, so this is encapsulated.

In V2 though, it is the responsibility of the caller to calculate it.

I agree that it is cleaner like that, I'll change it now.